server/src/main/java/org/opensearch/search/internal/ContextIndexSearcher.java

```diff
@@ -478,4 +485,38 @@ public void clear() {
             runnables.clear();
         }
     }
+
+    private boolean canMatch(LeafReaderContext ctx) throws IOException {
+        // skip segments for search after if min/max of them doesn't qualify competitive
+        return canMatchSearchAfter(ctx);
+    }
+
+    private boolean canMatchSearchAfter(LeafReaderContext ctx) throws IOException {
+        // Only applied on primary sort field and primary search_after.
+        FieldSortBuilder sortBuilder = searchContext == null
+            ? null
+            : FieldSortBuilder.getPrimaryFieldSortOrNull(searchContext.request().source());
+        MinAndMax<?> minMax = sortBuilder != null
+            ? FieldSortBuilder.getMinMaxOrNullForSegment(this.searchContext.getQueryShardContext(), ctx, sortBuilder)
+            : null;
+        return SearchService.canMatchSearchAfter(searchContext.searchAfter(), minMax, sortBuilder.order());
```

Comment:
We should better short-circuit here when `searchContext` or `sortBuilder` are `null`, oherwise NPEs guaranteed ...

Reply:
yup yup, just saw that :) correcting it.

Reply:
Done.